### PR TITLE
feat!: Update to ACVM v0.11.0

### LINKED ISSUE #159: Update to ACVM v0.11.0
### Problem

ACVM v0.11.0 changed the backend traits to allow failures by many functions needing to return Result. It also added keccak.

### Proposed solution

Update the project to v0.11.0 by updating the trait implementations and avoiding unwrap or panic in the codebase. Also, remove the duplicated keccak code which was added prior to 0.11 support.

### Alternatives considered

_No response_

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by phated**
This reworks the backend to leverage the changes for fallible functions. It introduces an error enum with admittedly bad variant names and uses them to map error cases throughout the codebase so panic and/or unwrap isn't used to handle most values.

~~There are still some TODOs to cleanup even more `unwrap` cases and it relies on something like https://github.com/noir-lang/acvm/pull/251, but this should be a solid skeleton for how a backend can surface errors to nargo.~~

Closes #159

### CODE REVIEW

**Comment by kevaundray** [comment]
You can change the commit to point to acvm 0.11 -- Is this ready for review? If not, then I can push a new PR with the fallible traits but has the same behaviour as without the fallible traits.

Main goal here is to get Keccak upstreamed -- Alternatively, I can add Keccak into the pwg for now by [copying this code](https://github.com/noir-lang/acvm/blob/443c73482eeef6cc42a1a254bf0d7706698ee353/acvm/src/pwg/hash.rs#L45) into [here](https://github.com/noir-lang/aztec_backend/blob/c9fb9e806f1400a2ff7594a0669bec56025220bb/src/acvm_interop/pwg.rs#L35). 

Then removing it once we use acvm > 0.11

**Comment by phated** [comment]
Promoting from draft because acvm 0.11 was released but we still need to workshop the errors and stuff.

**Comment by phated** [comment]
I've completed my `TryInto` TODOs and split the wasm errors into their own enum. `src/merkle.rs` still has a lot of unwraps but I don't think we need to update that for this to land.

Besides error variant naming and error messages, I think this should be good-to-go. Will wait on @TomAFrench to do a full review and give feedback on those.

**Review by TomAFrench — COMMENTED** [review]
A couple of these error types don't need to exist in my mind (comments below).

We've also bundled in a change which changes `WASMValue` to cover not just return values but all arguments as well. I'm going to split off this change so I can review errors in isolation.

File: src/schnorr.rs
Comment (phated) [reply src/schnorr.rs:27]: I'm confused. If this were handled by the type system, we wouldn't need to `try_into` and `unwrap`? Is there some other way to concat these that doesn't require the `try_into`?

Note that even the `some_usize_number as i32` syntax we use a lot isn't actually safe and will panic on some architectures. We should likely be using `try_into` on those too...

File: src/lib.rs
Comment (phated) [inline src/lib.rs:205]: I wonder if these should be From impls 🤔

File: src/schnorr.rs
Comment (TomAFrench) [reply src/schnorr.rs:27]: Well, it's _almost_ enforced. I guess what I mean is that correctness of this code is determined at compile time so we should try to avoid adding runtime checks on this.

We know that `construct_signature` will always return two 32 byte arrays and so we can concat them into a 64 byte array and that will work no matter what the inputs to the function are. We can make this explicit by adding a type declaration for the return values from `construct_signature` to be safer though.

File: src/schnorr.rs
Comment (TomAFrench) [reply src/schnorr.rs:100]: Ideally we'd add type aliases for the signatures etc, to express them in terms of `FIELD_BYTES`

File: src/composer.rs
Comment (TomAFrench) [inline src/composer.rs:66]: Can we assign the return value from `get_exact_circuit_size`to a variable and then return that as `Ok(circuit_size)`? Having an unsafe block nested like this feels iffy.

**Review by TomAFrench — APPROVED** [review]
I think that we can remove a number of errors which are included here. I don't think we should expose these as users never need to be concerned with them. I don't think it's worth blocking over however as we don't have semver set up yet so every release is breaking and we can just remove them progressively.

File: src/lib.rs
Comment (TomAFrench) [reply src/lib.rs:205]: Yeah, I don't have strong feelings either way on this. It's all internal to the crate so we can change this in a followup PR if necessary.

File: src/schnorr.rs
Comment (TomAFrench) [reply src/schnorr.rs:27]: I've made a PR https://github.com/noir-lang/aztec_backend/pull/163 which makes these conversions safer by centralising them into a single function which can be easily reasoned about.

File: src/merkle.rs
Comment (phated) [reply src/merkle.rs:68]: #166

File: src/lib.rs
Comment (phated) [reply src/lib.rs:95]: This doesn't seem possible because that's an enum and if I try something like:
```
#[error(transparent)]
UnsupportedBlackBoxFunc(#[from] OpcodeResolutionError::UnsupportedBlackBoxFunc()),
```
It says
> try using the variant's enum: `acvm::OpcodeResolutionError`, `crate::Error`, `crate::OpcodeResolutionError`

which is the incorrect way to handle this error since it can't be anything from the enum.

File: src/lib.rs
Comment (phated) [reply src/lib.rs:66]: This isn't possible with Associated Types. See this error:
Screenshot 2023-05-08 at 10 53 03 AM

I believe we'll need to make the nargo errors generic over the backend errors, but I've not finished that work yet.

File: src/composer.rs
Comment (phated) [reply src/composer.rs:66]: I'm pretty sure this is fine but I'll change it because the style doesn't matter to me

File: src/lib.rs
Comment (phated) [reply src/lib.rs]: I made them TryFrom and I like it a lot.

File: src/lib.rs
Comment (phated) [reply src/lib.rs:143]: I think we keep these until we fix them correctly via the compiler/type system. I'm noticing some stuff that needs to be changed in bb-sys from this work.

File: src/lib.rs
Comment (phated) [inline src/lib.rs:101]: In trying to make the `CliError` type generic over a Backend, I was getting an error that the backend didn't implement `Debug` so I derived it here.

This isn't ideal, since we don't require Debug for 0.11 but we may need to require it for 0.12. I'll link the Noir draft once it is done.

Comment (phated) [reply src/lib.rs:101]: See https://github.com/noir-lang/noir/pull/1322

File: src/lib.rs
Comment (TomAFrench) [reply src/lib.rs:64]: Have you just made the `Error` type private here? This isn't quite what I meant, instead I think we should wrap this in a struct which has what is now `Error` in a struct which will display the error message but doesn't allow matching on variants, etc.

**Comment by TomAFrench** [comment]
Some thoughts on error variants
 
 ```
// These variants shouldn't exist as we can prevent them at compile time
SchnorrConvert(Vec<u8>),
SchnorrSlice { source: TryFromSliceError },
FieldElementSlice { source: TryFromSliceError },
FieldToArray(usize, usize),

// Can be replaced with a MalformedOpcode(Opcode) error
// This error can be defined down in ACVM and be shared across backends
MissingOutput(usize),
MissingLeaf,
MissingLeafIndex,
MissingPublicKeyX,
MissingPublicKeyY,
MissingSignature(usize),
MissingPublicKey(&'static str, usize),

// I think we can leave this as a panic tbh.
// Calling a backend with an unsupported opcode is calling the `solve` function outside of contract.
Aes,

// Unsure what to do these
Pow2CeilOverflow(u32),
WasmError,
```

File: src/lib.rs
Comment (phated) [reply src/lib.rs:64]: Sounds good. I will push that shortly

File: src/acvm_interop/pwg.rs
Comment (TomAFrench) [inline src/acvm_interop/pwg.rs:108]: This error will never be thrown as it would require us to be using a 0 bit curve.

Comment (TomAFrench) [reply src/acvm_interop/pwg.rs:108]: We can also use a function such as https://github.com/noir-lang/acvm/pull/269/files#diff-c31b89a76cc86219fd7be804ffbb7dfe18d93b96924a0958f92a9b320572e2bbR6 here.

**Comment by phated** [comment]
I think this PR is good-to-go and we should hold any other changes/PRs until it lands. We need to get this stuff upstreamed.

**Review by kobyhallx — APPROVED** [review]
This is great! Thank You!

**Pull request merged**
